In this project, we analyse a dataset of used cars from the classified section of the German eBay website: eBay Kleinanzeigen. While the original dataset was obtained from Kaggle, it has been modified for this project in the following ways:

<ul>
<li> A sample of n = 50000 was retrieved from the full dataset.
<li> The dataset was slightly dirtied (original was cleaned)
</ul>

Our aim is to clean the dataset and then analyse the included used car listings.

The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.
name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
kilometer - How many kilometers the car has driven.
monthOfRegistration - The month in which the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.


In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [3]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

# Initial Observations of Dataset

We will need to clean the 'price' column by removing the '$' sign.

Under the notRepairedDamage column, there are both 'Nein' and 'NaN' values. We need to figure out what the 'NaN' values represent, as they could possibly also mean that the listing has no unrepaired damages.

We will need to clean the 'odometer' column by removing the 'km' string.


In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
mapping_dict = {
    'dateCrawled': 'date_crawled',
    'offerType': 'offer_type',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'powerPS': 'power_ps',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'nr_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen',
}

autos.rename(columns=mapping_dict, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We have modified the column names of our dataset and changed all the camelCase names to snakecase, in order to match Python's standard format.

In [7]:
autos.describe(include='all')


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The following columns have only 2 unique values out of 50000 counts, and can possibly be considered to have a homogeneous value throughout the dataset as one of the unique values has a very high frequency of 49999:
<ul>
<li> seller
<li> offer_type
</ul>

The following columns are missing data, as they have less values than the total number of 50000 counts:
<ul>
<li> vehicle_type
<li> gearbox
<li> model
<li> fuel_type
<li> unrepaired_damage
</ul>

The following columns contain numeric data which we can clean up and convert in numeric dtypes:
<ul>
<li> price
<li> odometer
</ul>

In [8]:
# Removing the 'seller' and 'offer_type' columns.

autos = autos.drop(columns=['seller', 'offer_type'])
autos.head(3)

date_crawled                                        name   price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  $5,000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  $8,500   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  $8,990   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   
2     test    limousine               2009    manuell       102    golf   

    odometer  registration_month fuel_type       brand unrepaired_damage  \
0  150,000km                   3       lpg     peugeot              nein   
1  150,000km                   6    benzin         bmw              nein   
2   70,000km                   7    benzin  volkswagen              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37

In [9]:
# Cleaning prices

price_column = autos.loc[:,'price']

def clean_prices(price_column):
    """function for cleaning the column of prices
    and converting it to float type"""
    
    price_column = price_column.str.replace(',', '')
    price_column = price_column.str.replace('$', '')
    price_column = price_column.astype(int)
    
    return price_column


modified_price_column = clean_prices(price_column)
modified_price_column
autos['price'] = modified_price_column


<ipython-input-9-59712a32d2b0>:10: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  price_column = price_column.str.replace('$', '')


In [10]:
# Cleaning odometer

odo_col = autos['odometer']


def clean_odo(col):
    """function to clean the odometer column
    and convert to int type"""
    col = col.str.replace(',', '')
    col = col.str.replace('km', '')
    col = col.astype(int)
    
    return col
    
    
modified_odo_col = clean_odo(odo_col)
autos['odometer'] = modified_odo_col

In [11]:
autos.rename(
    {'odometer': 'odometer_km',
     'price': 'price_in_dollars',
    }, inplace=True, axis=1)
autos.head(3)

date_crawled                                        name  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United   

   price_in_dollars   abtest vehicle_type  registration_year    gearbox  \
0              5000  control          bus               2004    manuell   
1              8500  control    limousine               1997  automatik   
2              8990     test    limousine               2009    manuell   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

We have now cleaned up both the 'price' and 'odometer' columns, and renamed them to 'price_in_dollars' and 'odometer_km' respectively. We proceed to check the data of these columns themselves, to look for any abnormal/outlying data points. We have also removed the non-significant columns of 'seller' and 'offer_type'.

In [12]:
print("The 'odometer_km' column's details are as follows: \n")
print(autos['odometer_km'].value_counts(), '\n')
print(autos['odometer_km'].describe())
print(autos['odometer_km'].unique().shape)


The 'odometer_km' column's details are as follows: 

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64 

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
(13,)


In [13]:
print("The 'price_in_dollars' column's details are as follows: \n")
print(autos['price_in_dollars'].value_counts().sort_index(ascending=True), '\n')
print(autos['price_in_dollars'].describe(), '\n')
print(autos['price_in_dollars'].unique().shape)


The 'price_in_dollars' column's details are as follows: 

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price_in_dollars, Length: 2357, dtype: int64 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_in_dollars, dtype: float64 

(2357,)


From looking at the details of the 'price_in_dollars' column, it is clear that there are some outliers in prices listed on a few ads. These include very low prices of < 100 dollars, to as high as 999999999 dollars, which is possibly the max price that can be listed on eBay. We let the cut-off "reasonable" low price be 1 dollar (as this is an auction site and it is common for people to set this to be the opening bid), and the cut-off "reasonable" high price to be 10000000 dollars. Below, we check which of the listings in our dataset fall OUTSIDE of this range.

In [14]:
price_bool = (autos['price_in_dollars'] < 1) | (autos['price_in_dollars'] >= 10000000)
autos.loc[price_bool, ['name', 'price_in_dollars']]


name  price_in_dollars
27                Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE                 0
71     Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...                 0
80     Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...                 0
87                         Bmw_520_e39_zum_ausschlachten                 0
99                       Peugeot_207_CC___Cabrio_Bj_2011                 0
...                                                  ...               ...
49884               Audi_a6_2.5l__Schnaeppchen_nur_heute                 0
49943                                         Opel_astra                 0
49960            Ford_KA_zu_verschenken_***Reserviert***                 0
49974  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...                 0
49984  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...                 0

[1429 rows x 2 columns]

Now, we remove these rows.

In [15]:
price_remove_bool = (autos['price_in_dollars'] >= 1) & (autos['price_in_dollars'] < 10000000)
autos = autos.loc[price_remove_bool, :]

In [16]:
print("The new 'price_in_dollars' column's details are as follows: \n")
print(autos['price_in_dollars'].value_counts().head(10), '\n')
print(autos['price_in_dollars'].describe(), '\n')
print(autos['price_in_dollars'].unique().shape)

The new 'price_in_dollars' column's details are as follows: 

500     781
1500    734
2500    643
1000    639
1200    639
600     531
800     498
3500    498
2000    460
999     434
Name: price_in_dollars, dtype: int64 

count    4.857100e+04
mean     6.082245e+03
std      2.278698e+04
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.890000e+06
Name: price_in_dollars, dtype: float64 

(2351,)


The most common prices are around the 500 to 4000 dollars range, with the median being 3000 dollars.

We now take a look at the 'date_crawled', 'ad_created', 'last_seen', and 'registration_year' columns, and attempt to calculate the distributions of dates for each one of them.


In [17]:
# details of the various date-related columns

autos.loc[:, ['date_crawled', 'ad_created', 'last_seen', 'registration_year']].head(10)


date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21   
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59   
7  2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32   
8  2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10   
9  2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35   

   registration_year  
0               2004  
1               1997  
2               2009  
3               2007  
4               2003  
5               2006  
6               1995  
7               1998  
8               2000  
9               1997

From the above, it is clear that for the first 3 columns, the date consists of the first 10 characters of each entry strings.
For the registration_year, the dtype is int, so we can work with it directly.

In [18]:
# 'date_crawled' column

date_crawled = autos['date_crawled'].str[:4]
date_crawled.value_counts(normalize=True, dropna=False)

2016    1.0
Name: date_crawled, dtype: float64

All the data captured in this set were collected in 2016.

In [19]:
# 'ad_created' column

ad_created = autos['ad_created'].str[:4]
ad_created.value_counts(normalize=True, dropna=False)

2016    0.999876
2015    0.000124
Name: ad_created, dtype: float64

Virtually all posts that were available in 2016 were made in 2016, with only a tiny group being from 2015. No posts from before that were up on the site in 2016.

In [20]:
# 'last_seen' column

last_seen = autos['last_seen'].str[:4]
last_seen.value_counts(normalize=True, dropna=False)



2016    1.0
Name: last_seen, dtype: float64

All posts had last been seen by the crawler in 2016.

In [21]:
# 'registration_year' column

registration_year = autos['registration_year']
print(registration_year.describe(), '\n')
print(registration_year.astype(str).describe(), '\n')
print(registration_year.value_counts(normalize=True, dropna=False).sort_index(ascending=True))

count    48571.000000
mean      2004.754257
std         88.638548
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64 

count     48571
unique       95
top        2000
freq       3156
Name: registration_year, dtype: object 

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64


The most common year cars were registered in was 2000.
We have weird years such as 4500, 4100, 5911, 9000, 1000 etc., which are impossible years as they are either too far into the future, or before cars were even invented.
We set the cutoff lowest year to be 1956, as according to this [link](http://www.hegis.de/fed__rep__of_germany.htm), the current registration system was set in stone in this year. The cutoff highest year is 2016, as that is the year this dataset was collected, and there is no way a car can be registered in a year that was after it and still be listed.

In [22]:
# removing entries outside of the 1956 - 2016 range

reg_bool = (autos['registration_year'] >= 1956) & (autos['registration_year'] <= 2016)
autos = autos.loc[reg_bool, :]


In [23]:
# 'registration_year' column

registration_year = autos['registration_year']
print(registration_year.describe(), '\n')
print(registration_year.astype(str).describe(), '\n')
print(registration_year.value_counts(normalize=True, dropna=False).head(10))


count    46656.000000
mean      2002.952911
std          6.980751
min       1956.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64 

count     46656
unique       61
top        2000
freq       3156
Name: registration_year, dtype: object 

2000    0.067644
2005    0.062929
1999    0.062114
2004    0.057935
2003    0.057849
2006    0.057249
2001    0.056499
2002    0.053284
1998    0.050647
2007    0.048804
Name: registration_year, dtype: float64


All the remaining entries now have registration years that are within a reasonable range. The most common registration years are still around the 1998 - 2006 years, with 2000 still being the most common.

We now use aggregation techniques to explore the 'brand' column. To do so, we first identify the unique values of the column, and then create an empty dictionary to store our aggregate data.


In [24]:
# finding the unique values in the 'brand' column

brand_col = autos['brand']

brand_unique = brand_col.unique()
print(brand_unique)
print(brand_col.value_counts(normalize=True, dropna=False))

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']
volkswagen        0.211398
bmw               0.110104
opel              0.107575
mercedes_benz     0.096429
audi              0.086613
ford              0.069873
renault           0.047154
peugeot           0.029857
fiat              0.025656
seat              0.018261
skoda             0.016418
nissan            0.015282
mazda             0.015196
smart             0.014168
citroen           0.013996
toyota            0.012710
hyundai           0.010031
sonstige_autos    0.009645
volvo             0.009152
mini              0.008766
mitsubishi        0.008230
honda             0.007845
kia               0.007073

From the above, we decide to only analyse brands with a percentage of 1% or higher of the total counts i.e. hyundai and above, as that is how I define a significant proportion of the population.

In [25]:
# removing all values with less than 1% proportion of population

print(type(brand_col.value_counts(normalize=True, dropna=False)))

brand_counts = brand_col.value_counts(normalize=True, dropna=False)
brand_sig_bool = (brand_counts >= 0.01)
brand_sig = brand_counts.loc[brand_sig_bool] # significant 'brand' values only left

print(brand_sig)

brand_sig_list = brand_sig.index.to_list()

<class 'pandas.core.series.Series'>
volkswagen       0.211398
bmw              0.110104
opel             0.107575
mercedes_benz    0.096429
audi             0.086613
ford             0.069873
renault          0.047154
peugeot          0.029857
fiat             0.025656
seat             0.018261
skoda            0.016418
nissan           0.015282
mazda            0.015196
smart            0.014168
citroen          0.013996
toyota           0.012710
hyundai          0.010031
Name: brand, dtype: float64


We now proceed to create the aggregate dictionary

In [40]:
# creating aggregate dictionary for 'brand'

brands_mean_price_dict = {}

for brand in brand_sig_list:
    brand_bool = (autos['brand'] == brand)
    brand_filtered = autos.loc[brand_bool, :]
    mean_price = brand_filtered.loc[:, 'price_in_dollars'].mean()
    brands_mean_price_dict[brand] = mean_price

print('The average prices for each brand are: \n')
print(brands_mean_price_dict)

print('\nA summary of the average prices are as follows: \n')
print('Max average price is', max(brands_mean_price_dict.values()))
print('Minimum average price is', min(brands_mean_price_dict.values()))

print('\nIn ranked order, here are all the average prices: \n')

new_dict, avg_prices = {}, []

for key in brands_mean_price_dict:
    new_dict[brands_mean_price_dict[key]] = key
    avg_prices.append(brands_mean_price_dict[key])

avg_prices.sort(reverse=True)
for price in avg_prices:
    print(new_dict[price], price)

    

The average prices for each brand are: 

{'volkswagen': 5600.391260265639, 'bmw': 8570.910064239828, 'opel': 2976.2472604104405, 'mercedes_benz': 8580.198266281395, 'audi': 9336.687453600594, 'ford': 4025.9257668711657, 'renault': 2475.7172727272728, 'peugeot': 3094.0172290021537, 'fiat': 2813.748538011696, 'seat': 4402.389671361502, 'skoda': 6368.0, 'nissan': 4743.40252454418, 'mazda': 4112.596614950635, 'smart': 3580.2239031770046, 'citroen': 3761.957120980092, 'toyota': 5167.091062394604, 'hyundai': 5365.254273504273}

A summary of the average prices are as follows: 

Max average price is 9336.687453600594
Minimum average price is 2475.7172727272728

In ranked order, here are all the average prices: 

audi 9336.687453600594
mercedes_benz 8580.198266281395
bmw 8570.910064239828
skoda 6368.0
volkswagen 5600.391260265639
hyundai 5365.254273504273
toyota 5167.091062394604
nissan 4743.40252454418
seat 4402.389671361502
mazda 4112.596614950635
ford 4025.9257668711657
citroen 3761.95712098

From the above, we see that the brand with the highest average price is 'audi', at 9336.68 dollars. The brand with the lowest average price is 'renault', at 2475.71 dollars.

From the ranked list above, we see that audi, mercedes_benz, and bmw average prices have a big jump from the next few brands (skoda, volkswagen, hyundai, etc.)

With the above information, we now seek trends/relations between the other attributes of a car and its selling price. We specifically focus on the average mileage for each of the brands listed above, and compare them with their average selling prices. We do so by constructing a new aggregate dictionary that has the average odometer values, and then use panda's series and dataframe constructors to create a dataframe with the 2 columns, indexed by the brand.

In [42]:
# creating aggregate dictionary for 'odometer'

brands_avg_odometer_dict = {}

for brand in brand_sig_list:
    brand_bool = (autos['brand'] == brand)
    brand_filtered = autos.loc[brand_bool, :]
    mean_odometer = brand_filtered.loc[:, 'odometer_km'].mean()
    brands_avg_odometer_dict[brand] = mean_odometer

print(brands_avg_odometer_dict)

{'volkswagen': 128714.38710331543, 'bmw': 132600.73973136072, 'opel': 129336.5212193664, 'mercedes_benz': 130865.74794398755, 'audi': 129157.38678544914, 'ford': 124391.10429447853, 'renault': 128127.27272727272, 'peugeot': 127153.62526920316, 'fiat': 117121.9715956558, 'seat': 121267.60563380281, 'skoda': 110848.5639686684, 'nissan': 118330.99579242637, 'mazda': 124464.03385049365, 'smart': 99326.77760968229, 'citroen': 119724.34915773354, 'toyota': 115944.35075885328, 'hyundai': 106442.30769230769}


In [46]:
# using series constructor to create mean_price and mean_odometer series

mean_price_series = pd.Series(brands_mean_price_dict)
mean_odo_series = pd.Series(brands_avg_odometer_dict)


In [50]:
# using dataframe constructor to create dataframe with 2 columns

price_vs_odo_df = pd.DataFrame(mean_price_series, columns=['mean_price'])
price_vs_odo_df['mean_odometer'] = mean_odo_series
print(price_vs_odo_df)


                mean_price  mean_odometer
volkswagen     5600.391260  128714.387103
bmw            8570.910064  132600.739731
opel           2976.247260  129336.521219
mercedes_benz  8580.198266  130865.747944
audi           9336.687454  129157.386785
ford           4025.925767  124391.104294
renault        2475.717273  128127.272727
peugeot        3094.017229  127153.625269
fiat           2813.748538  117121.971596
seat           4402.389671  121267.605634
skoda          6368.000000  110848.563969
nissan         4743.402525  118330.995792
mazda          4112.596615  124464.033850
smart          3580.223903   99326.777610
citroen        3761.957121  119724.349158
toyota         5167.091062  115944.350759
hyundai        5365.254274  106442.307692


From the data above, we see that despite having approximately the same mean odometer mileages, volkswagen, bmw, and audi have much higher mean prices than other brands such as renault, fiat, opel, etc. 

This possibly indicates that the brand itself is an influencing factor on the price of its listing, as the highly priced brands are also typically considered to be 'luxury' brands.